Observer/ImportRates.php

```diff
@@ -278,80 +281,134 @@ private function setZipCode(string $value): self
     }
 
     /**
-     * @param Observer $observer
-     * @return self
      * @throws LocalizedException
      */
-    public function execute(Observer $observer): self
+    public function execute(Observer $observer)
     {
-        if ($this->backupRatesEnabled() && $this->taxjarConfig->getApiKey()) {
-            $client = $this->clientFactory->create();
-            $zipCode = $this->backupRateOriginAddress->getShippingZipCode();
-            $customerTaxClassConfig = $this->scopeConfig->getValue(TaxjarConfig::TAXJAR_CUSTOMER_TAX_CLASSES);
-            $productTaxClassConfig = $this->scopeConfig->getValue(TaxjarConfig::TAXJAR_PRODUCT_TAX_CLASSES);
-            $shippingTaxClass = $this->scopeConfig->getValue(MagentoTaxConfig::CONFIG_XML_PATH_SHIPPING_TAX_CLASS);
-
-            $this->setClient($client)
-                ->setZipCode($zipCode)
-                ->setCustomerTaxClasses($customerTaxClassConfig)
-                ->setProductTaxClasses($productTaxClassConfig)
-                ->setShippingTaxClass($shippingTaxClass)
-                ->importRates();
-        } else {
-            $this->purgeExistingRates();
-            $this->setLastUpdate(null);
-            $this->resourceConfig->saveConfig(TaxjarConfig::TAXJAR_BACKUP, 0, 'default');
-            $this->messageManager->addNoticeMessage(__('Backup rates imported by TaxJar have been queued for removal.'));
+        $updated = null;
+        $rates = $codes = [];
+
+        $isActive = $this->backupRatesEnabled() && $this->taxjarConfig->getApiKey();
+
+        if ($isActive) {
+            $this->setConfiguration();
+            $this->validateConfiguration();
+
+            $rates = $this->getRates();
+            $updated = $this->getDate();
         }
 
-        $this->cacheManager->flush($this->cacheManager->getAvailableTypes());
+        $rateCount = count($rates);
 
-        return $this;
+        if ($rateCount) {
+            foreach($rates as $rate) {
+                $codes[] = self::getRateCode($rate);
+            }
+        }
+
+        if ($this->debugEnabled()) {
+            $this->messageManager->addNoticeMessage(
+                __('Debug mode enabled. Backup tax rates have not been altered.')
+            );
+            return;
+        }
+
+        $this->upsertRates($rates);
+        $this->purgeRates($codes);
+        $this->setRateCount($rateCount);
+        $this->setLastUpdate($updated);
+        $this->resourceConfig->saveConfig(TaxjarConfig::TAXJAR_BACKUP, (int) $isActive);
```

Comment:
Why do we need to update this configuration here?

Reply:
In the "else" statement on the old version of this diff, this config value was set equal to false. Since the if/else was refactored away, I retained that functionality by including this line.. Although it was not apparent to me "why" we would want to do that.

It seems like the only way that it would really have any effect would be if backup rates were enabled, then the api key was removed or the extension was disabled.

I didn't see any harm in calling update each time, but if we only wanted to execute the line in the case mentioned above, we could wrap it in an if like so:
```
if (! $isActive) {
    $this->resourceConfig->saveConfig(TaxjarConfig::TAXJAR_BACKUP, 0);
}
```